File: array/48.py
from typing import List
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # transpose
        N = len(matrix)
        for i in range(N):
            for j in range(0, i):
                matrix[i][j] = matrix[i][j] ^ matrix[j][i]
                matrix[j][i] = matrix[i][j] ^ matrix[j][i]
                matrix[i][j] = matrix[i][j] ^ matrix[j][i]
        
        # print(matrix)
        # 左右互换
        for i in range(N):
            for j in range(N // 2):
                matrix[i][j] = matrix[i][j] ^ matrix[i][N - j - 1]
                matrix[i][N - j - 1] = matrix[i][j] ^ matrix[i][N - j - 1]
                matrix[i][j] = matrix[i][j] ^ matrix[i][N - j - 1]
        # print(matrix)
def main():
    matrix = [
  [ 5, 1, 9,11],
  [ 2, 4, 8,10],
  [13, 3, 6, 7],
  [15,14,12,16]
    ]
    Solution().rotate(matrix)
    # print(matrix)

if __name__ == "__main__":
    main()

File: leetcode-cpp/48.cpp
#include <iostream>
#include <algorithm>

using namespace std;

class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        if (matrix.size() == 0 || matrix.size() == 1) return;
        // transpose
        int s = matrix.size();
        for (int i = 0; i < s; i++){
            for (int j = i + 1; j < s; j++){
                matrix[i][j] = matrix[i][j] ^ matrix[j][i];
                matrix[j][i] = matrix[i][j] ^ matrix[j][i];
                matrix[i][j] = matrix[i][j] ^ matrix[j][i];
            }
        }
        // reverse
        for (int i = 0; i < s; i++) {
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};

int main(){
    Solution s;
    vector<vector<int>> ret = {
    {1, 2, 3},
    {4, 5, 6},
    {7, 8, 9},
    };
    s.rotate(ret);
    for (int i = 0; i < ret.size(); i++){
        for (int j = 0; j < ret[i].size(); j++){
            cout << ret[i][j] << "\t";
        }
        cout <<endl;
    }
}
